### PR TITLE
Features/bilstm

### PR DESCRIPTION
**Pull request by ArturoSanchezPalacio**
Regression to estimate numerical factors (TAS20, F1, F2 or F3). Added context to this model.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+[packages]
+flask = "*"
+tornado = "*"
+pillow = "*"
+
+
+tensorflow = "*"
+keras ="*"
+[requires]
+python_version = "3.6"
```
Comment (jorgemf) [inline Pipfile]: can we use python 3.7?

File: polyaxon/notebook.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+kind: notebook
+
+environment:
+  resources:
+    cpu:
+      requests: 4
+      limits: 4
+    memory:
+      requests: 16000
```
Comment (jorgemf) [inline polyaxon/notebook.yml]: can we decrease this, so we can launch more than one notebook in the machine (like 10000 for requests and 12000 for limits, or even less for both)

File: polyaxon/one_gpu_classification.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+build:
+  image: tensorflow/tensorflow:1.14.0-gpu-py3
+  build_steps:
+    - pip install --no-cache-dir -U polyaxon-cli
+    - pip install --no-cache-dir pandas
+    - pip install --no-cache-dir keras
+    - pip install --no-cache-dir sklearn
+
+params:
```
Comment (jorgemf) [inline polyaxon/one_gpu_classification.yml:27]: These parameters are to display them in polyaxon. Probably is not worth to have the input path and the file names in this section. It will add noise to polyaxon. Furthermore, this parameters are suppose to change for the experiments, no to be constant.

File: polyaxon/one_gpu_regression.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+build:
+  image: tensorflow/tensorflow:1.14.0-gpu-py3
+  build_steps:
+    - pip install --no-cache-dir -U polyaxon-cli
+    - pip install --no-cache-dir pandas
+    - pip install --no-cache-dir keras
+    - pip install --no-cache-dir sklearn
+
+params:
+  input_path: /data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics
```
Comment (jorgemf) [inline polyaxon/one_gpu_regression.yml]: Same with this paramterers

File: polyaxon/two_gpu.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+---
```
Comment (jorgemf) [inline polyaxon/two_gpu.yml]: You can remove the 2 GPU version for this project

File: src/bilstm_train_classification.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+from sklearn.metrics import roc_auc_score
+from sklearn.metrics import confusion_matrix
+
+
+
+
+logger = logging.getLogger(__name__)
+
+
+def setup_script(filename_log=None):
```
Comment (jorgemf) [inline src/bilstm_train_classification.py:29]: This method is never called. For me all the content inside the method should execute after the line 33. 

It happens the same with the other file

File: src/bilstm_train_regression.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import roc_auc_score
+
+
+
+
+logger = logging.getLogger(__name__)
+
+
+def setup_script(filename_log=None):
```
Comment (jorgemf) [inline src/bilstm_train_regression.py:31]: Never executed. Already commented in the other file

File: src/bilstm_train_classification.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        datefmt=datefmt,
+    )
+    if filename_log:
+        build_log = logging.FileHandler(filename=filename_log, mode='w')
+        build_log.setLevel(logging.INFO)
+        formatter = logging.Formatter(fmt, datefmt=datefmt)
+        build_log.setFormatter(formatter)
+        logging.getLogger('').addHandler(build_log)
+
+class Configuration(argparse.Namespace):
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: This code is duplicated in 2 files, it should be in a separeted file and imported

File: src/bilstm_train_classification.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    def __delitem__(self, name):
+        raise Exception('Params cannot be deleted')
+
+    def copy(self, **kwargs):
+        new_params = self.__dict__.copy()
+        for key, value in kwargs.iteritems():
+            new_params[key] = value
+        return Configuration(**new_params)
+
+def get_args():
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: This code is very probably duplicated. Better in a separeted file

File: src/bilstm_train_classification.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    args_vars_list = []
+    for k in sorted(args_vars.keys()):
+        if not (isinstance(args_vars[k], types.FunctionType) or k.startswith('__')):
+            args_vars_list.append('{}: {}'.format(k, args_vars[k]))
+    logging.info('Using command args: { %s }', ', '.join(args_vars_list))
+    polyaxon_utils.set_params(args_vars)
+
+    return args
+
+def get_tokenizer(args):
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: This is also probably duplicated

File: src/bilstm_train_classification.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    x = SpatialDropout1D(0.2)(x)
+    x = Bidirectional(GRU(128, return_sequences=True, dropout=0.1, recurrent_dropout=0.1))(x)
+    x = Conv1D(64, kernel_size=3, padding="valid", kernel_initializer="glorot_uniform")(x)
+    avg_pool = GlobalAveragePooling1D()(x)
+    max_pool = GlobalMaxPooling1D()(x)
+    x = concatenate([avg_pool, max_pool])
+    preds = Dense(1, activation="sigmoid")(x)
+    model = Model(sequence_input, preds)
+    model.compile(loss='binary_crossentropy', optimizer=Adam(lr=args.learning_rate), metrics=['accuracy'])
+    #model.compile(loss='mse', optimizer=Adam(lr=1e-3), metrics=['accuracy'])
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: too many empty lines and comments (nothing important, but just make the code less readeable)

File: src/bilstm_train_classification.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        if i >= args.max_features:
+            continue
+        embedding_vector = embeddings_index.get(word)
+        if embedding_vector is not None:
+            # words not found in embedding index will be all-zeros.
+            embedding_matrix[i] = embedding_vector
+
+    sequence_input = Input(shape=(args.max_length, ))
+    x = Embedding(args.max_features, args.embed_size, weights=[embedding_matrix], trainable=False)(sequence_input)
+    x = SpatialDropout1D(0.2)(x)
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: This is not needed as the dropout is also in the next line.

File: src/bilstm_train_classification.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+    max_pool = GlobalMaxPooling1D()(x)
+    x = concatenate([avg_pool, max_pool])
+    preds = Dense(1, activation="sigmoid")(x)
+    model = Model(sequence_input, preds)
+    model.compile(loss='binary_crossentropy', optimizer=Adam(lr=args.learning_rate), metrics=['accuracy'])
+    #model.compile(loss='mse', optimizer=Adam(lr=1e-3), metrics=['accuracy'])
+
+
+
+    X_tra, X_val, y_tra, y_val = train_test_split(x_train, y_train, train_size=0.9, random_state=233)
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: why this if we already have a test set in the arguments of the program? I think the other is legacy and shouldn't be there

File: src/bilstm_train_classification.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        super(Callback, self).__init__()
+
+        self.interval = interval
+        self.X_val, self.y_val = validation_data
+
+    def on_epoch_end(self, epoch, logs={}):
+        if epoch % self.interval == 0:
+            y_pred = self.model.predict(self.X_val, verbose=0)
+            score = roc_auc_score(self.y_val, y_pred)
+            print("\n ROC-AUC - epoch: {:d} - score: {:.6f}".format(epoch+1, score))
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: change to use looger

File: src/bilstm_train_classification.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    checkpoint = ModelCheckpoint(os.path.join(args.outputs_dir, args.weights_file), monitor='val_acc', verbose=1, save_best_only=True, mode='max')
+    early = EarlyStopping(monitor="val_acc", mode="max", patience=args.patience)
+    ra_val = RocAucEvaluation(validation_data=(X_val, y_val), interval=1)
+    callbacks_list = [ra_val, checkpoint, early]
+    model.fit(X_tra, y_tra, batch_size=args.batch_size, epochs=args.epochs, validation_data=(X_val, y_val),
+              callbacks=callbacks_list, verbose=1)
+
+    y_pred = model.predict(x_test, batch_size=1024, verbose=1)
+    score = roc_auc_score(y_test, y_pred)
+    print(y_test)
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: change to use looger, we shouldn't have any print if we have a logger

File: src/bilstm_train_regression.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+def get_tokenizer(args):
+    if args.read_tokenizer:
+        with open(os.path.join(args.models_path, args.tokenizer_file), 'rb') as handle:
+            tokenizer = pickle.load(handle)
+    else:
+        tokenizer = text.Tokenizer(num_words=args.max_features, lower=True)
+    return tokenizer
+
+
+class RocAucEvaluation(Callback):
```
Comment (jorgemf) [inline src/bilstm_train_regression.py]: this doesn't make sense for regression

File: src/bilstm_train_regression.py [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+        if i >= args.max_features:
+            continue
+        embedding_vector = embeddings_index.get(word)
+        if embedding_vector is not None:
+            # words not found in embedding index will be all-zeros.
+            embedding_matrix[i] = embedding_vector
+
+    text_input = Input(shape=(args.max_length,), name='text_input')
+    x = Embedding(args.max_features, args.embed_size, weights=[embedding_matrix], trainable=False)(text_input)
+    x = SpatialDropout1D(0.2)(x)
```
Comment (jorgemf) [inline src/bilstm_train_regression.py]: remove this

File: src/bilstm_train_regression.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+    # Adding context
+    auxiliary_input = Input(shape=(len(cards_train.columns),), name='aux_input')
+    x_1 = RepeatVector(args.max_length)(auxiliary_input)
+    x = keras.layers.concatenate([x, x_1])
+
+    x = Bidirectional(GRU(128, return_sequences=True, dropout=0.1, recurrent_dropout=0.1))(x)
+    x = Conv1D(64, kernel_size=3, padding="valid", kernel_initializer="glorot_uniform")(x)
+    avg_pool = GlobalAveragePooling1D()(x)
+    max_pool = GlobalMaxPooling1D()(x)
+    x = concatenate([avg_pool, max_pool])
```
Comment (jorgemf) [inline src/bilstm_train_regression.py]: It is better to chose either objects of functions instead of mixing them both in the same program. I would prefer the object concatenate here as the rest of the elements are objects. It is easier to understand if everything is consistent.

**Review by jorgemf — CHANGES_REQUESTED** [review]

File: src/bilstm_train_regression.py (hunk 17 above)
Comment (ArturoSanchezPalacio) [reply src/bilstm_train_regression.py]: What do you mean? That's the model we are using

Comment (jorgemf) [reply src/bilstm_train_regression.py]: The Line: ` x = SpatialDropout1D(0.2)(x)` That dropout doesn't make a lot of sense like in the other model

File: src/bilstm_train_classification.py (hunk 13 above)
Comment (ArturoSanchezPalacio) [reply src/bilstm_train_classification.py]: I think Javi uses this because he has lots of data so he can see when the network overfits and then validate with the test set. However for this project is not useful so I will erase it.

File: polyaxon/one_gpu_classification.yml [hunk 19]
```diff
@@ -25,17 +25,14 @@ build:
...
-  test_filename: test.csv
   models_path: /data/models/prolexitim_alexitimia
   tokenizer_file: tokenizer.pk
   embedding_path: /data/models/toxic_speech
   embedding_file: SBW-vectors-300-min5.txt
   embedding_matrix_file: embedding_matrix.pk
   weights_file: weights_base.best.hdf5
   max_features: 1000
-  max_length: 150
+  max_length: 284
```
Comment (jorgemf) [inline polyaxon/one_gpu_classification.yml]: is this the sequence max length?

File: polyaxon/one_gpu_regression.yml [hunk 20]
```diff
@@ -37,7 +34,7 @@ params:
   max_features: 1000
   max_length: 284
   embed_size: 300
-  epochs: 3
+  epochs: 1
```
Comment (jorgemf) [inline polyaxon/one_gpu_regression.yml]: increase the epochs

File: polyaxon/one_gpu_classification.yml [hunk 21]
```diff
@@ -45,4 +42,4 @@ params:
 
 run:
   cmd:
-    - python src/bilstm_train_classification.py --input_path {{ input_path }} --train_filename {{ train_filename }} --test_filename {{ test_filename }} --models_path {{ models_path }} --tokenizer_file {{ tokenizer_file }} --embedding_path {{ embedding_path }} --embedding_file {{ embedding_file }} --embedding_matrix_file {{ embedding_matrix_file }} --weights_file {{ weights_file }} --max_features {{ max_features }} --max_length {{ max_length }}  --embed_size {{ embed_size }}  --epochs {{ epochs }}  --batch_size {{ batch_size }}  --learning_rate {{ learning_rate }} --patience {{ patience }}
\ No newline at end of file
+    - python src/bilstm_train_classification.py  --models_path {{ models_path }} --tokenizer_file {{ tokenizer_file }} --embedding_path {{ embedding_path }} --embedding_file {{ embedding_file }} --embedding_matrix_file {{ embedding_matrix_file }} --weights_file {{ weights_file }} --max_features {{ max_features }} --max_length {{ max_length }}  --embed_size {{ embed_size }}  --epochs {{ epochs }}  --batch_size {{ batch_size }}  --learning_rate {{ learning_rate }} --patience {{ patience }}
```
Comment (jorgemf) [inline polyaxon/one_gpu_classification.yml]: I meant: `python src/bilstm_train_classification.py --input_path /data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics ...` instead of `python src/bilstm_train_classification.py --input_path {{ input_path }}`

File: src/bilstm_train_classification.py [hunk 22]
```diff
@@ -50,157 +53,36 @@ def setup_script(filename_log=None):
...
-
 
 if __name__ == '__main__':
 
     args = get_args()
     logger.info(f"Reading test file")
-    test = process_data_classification(args.input_path, args.test_filename)
-    X_test = test["Text-EN"].str.lower()
-    y_test = test['AlexLabel'].values
+    test_file = "/data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics/test.csv"
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: this should be as it was before

File: src/bilstm_train_classification.py [hunk 23]
```diff
@@ -50,157 +53,36 @@ def setup_script(filename_log=None):
...
+    texts_test = X_test["Text-EN"].str.lower()
+    cards_test = X_test.filter(regex="card_.*")  # Using a regular expression cause some cards may not be in test.
+    y_test = test['AlexLabel']
     logger.info(f" Using test dataset with {X_test.shape[0]} instances")
 
     logger.info(f"Reading train file")
-    train = process_data_classification(args.input_path, args.train_filename)
-    X_train = train["Text-EN"].str.lower()
-    y_train = train['AlexLabel'].values
+    train_file = "/data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics/train.csv"
```
Comment (jorgemf) [inline src/bilstm_train_classification.py]: this should be as it was before

File: src/bilstm_train_classification.py [hunk 24]
```diff
@@ -222,34 +104,29 @@ def on_epoch_end(self, epoch, logs={}):
...
-    y_pred = model.predict(x_test, batch_size=1024, verbose=1)
+    y_pred = model.predict([x_test, cards_test], batch_size=1024, verbose=1)
     score = roc_auc_score(y_test, y_pred)
-    print(y_test)
-    print(y_pred)
-    #cm = confusion_matrix(y_test, y_pred)
     logger.info(f"AUC for test {score}")
-    logger.info(f"Confusion Matrix: {cm}")
-    print("AUC for test",score)
+    logger.info("Saving model")
```
Comment (jorgemf) [inline src/bilstm_train_classification.py:126]: IT would be nice to keep the confusion matrix here

File: src/bilstm_train_regression.py [hunk 25]
```diff
@@ -49,151 +53,22 @@ def setup_script(filename_log=None):
...
-            score = roc_auc_score(self.y_val, y_pred)
-            print("\n ROC-AUC - epoch: {:d} - score: {:.6f}".format(epoch+1, score))
-
 
 if __name__ == '__main__':
 
     args = get_args()
     logger.info(f"Reading test file")
-    test = process_data_regression(args.input_path, args.test_filename)
+    test_file = "/data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics/test.csv"
```
Comment (jorgemf) [inline src/bilstm_train_regression.py]: this should be as before

File: src/bilstm_train_regression.py [hunk 26]
```diff
@@ -49,151 +53,22 @@ def setup_script(filename_log=None):
...
+    test = process_data_regression(test_file)
     X_test = test.drop(args.factor, axis=1)
     texts_test = X_test["Text-EN"].str.lower()
     cards_test = X_test.filter(regex="card_.*")  # Using a regular expression cause some cards may not be in test.
     y_test = test[args.factor]
     logger.info(f" Using test dataset with {X_test.shape[0]} instances")
 
     logger.info(f"Reading train file")
-    train = process_data_regression(args.input_path, args.train_filename)
+    train_file = "/data/serendeepia/prolexitim_alexitimia/20190920_prolexytim_analytics/train.csv"
```
Comment (jorgemf) [inline src/bilstm_train_regression.py]: this should be as before

**Review by jorgemf — CHANGES_REQUESTED** [review]

File: src/bilstm_train_classification.py [hunk 27]
```diff
@@ -116,6 +118,8 @@ def setup_script(filename_log=None):
 
     y_pred = model.predict([x_test, cards_test], batch_size=1024, verbose=1)
     score = roc_auc_score(y_test, y_pred)
+    y_pred_classes = [1 * (x[0]>=0.5) for x in y_pred]
+    cm = confusion_matrix(y_test, y_pred_classes)
```
Comment (jorgemf) [inline src/bilstm_train_classification.py:122]: print it in the logger

**Review by jorgemf — APPROVED** [review]
